File: day09/sensor.lua
local file = io.open('sensor.txt', 'r')
local totalExtrapolatedValues = 0
local totalExtrapolatedValuesBackwards = 0

for line in file:lines() do
    local sequence = {}
    for num in string.gmatch(line, "%S+") do
        table.insert(sequence, tonumber(num))
    end
    local sequences = {sequence}

    while #sequence > 0 and sequence[1] ~= 0 do
        local diffSequence = {}
        for i = 1, #sequence - 1 do
            table.insert(diffSequence, sequence[i + 1] - sequence[i])
        end
        table.insert(sequences, diffSequence)
        sequence = diffSequence
    end

    table.insert(sequences[#sequences], 1, 0)
    table.insert(sequences[#sequences], 0)

    for i = #sequences - 1, 1, -1 do
        table.insert(sequences[i], 1, sequences[i][1] - sequences[i + 1][1])
        table.insert(sequences[i], sequences[i][#sequences[i]] + sequences[i + 1][#sequences[i + 1]])
    end

    totalExtrapolatedValuesBackwards = totalExtrapolatedValuesBackwards + sequences[1][1]
    totalExtrapolatedValues = totalExtrapolatedValues + sequences[1][#sequences[1]]
end

print(totalExtrapolatedValues)
print(totalExtrapolatedValuesBackwards)

file:close()

File: day09/sensor.py
with open('sensor.txt') as f:
    sum_of_extrapolated_values = 0
    sum_of_extrapolated_values_backwards = 0

    # read the input line by line
    for line in f.readlines(): 
        sequence = list(map(int, line.split()))
        sequences = [sequence]

        # calculate diff sequences
        while not all(v == 0 for v in sequence):
            diff_sequence = [(sequence[i + 1] - sequence[i]) for i in range(len(sequence) - 1)]
            sequences.append(diff_sequence)
            sequence = diff_sequence

        # extrapolate previous/next value in last sequence (sequence of all zeroes)
        sequences[-1].insert(0, 0)
        sequences[-1].append(0)

        # extrapolate previous/next value for the other sequences in reverse order
        for i in reversed(range(len(sequences) - 1)):
            sequences[i].insert(0, sequences[i][0] - sequences[i + 1][0])
            sequences[i].append(sequences[i][-1] + sequences[i + 1][-1])

        # add extrapolated values to the original sequence
        sum_of_extrapolated_values_backwards += sequences[0][0]
        sum_of_extrapolated_values += sequences[0][-1]

    print(sum_of_extrapolated_values)
    print(sum_of_extrapolated_values_backwards)